src/librustc/hir/map/mod.rs

```diff
@@ -737,6 +737,29 @@ impl<'hir> Map<'hir> {
         }, |_| false).ok()
     }
 
+    /// Returns the outermost existential type node, given the node-ID of an existential type node
+    /// that is possibly nested in others.
+    pub fn get_outermost_existential(&self, id: NodeId) -> NodeId {
+        let mut cur_id = id;
+        loop {
+            let parent_id = self.get_parent_node(cur_id);
+            if parent_id == CRATE_NODE_ID {
+                break;
+            }
+            match self.get(parent_id) {
+                Node::Item(i) => {
+                    if let ItemKind::Existential(..) = i.node {
```

Comment:
Unrelated to any existing problems:

You should break at the first `ItemKind::Existential` whose `impl_trait_fn` is `None`, not on the outermost one. It's certainly possible to declare explicit existential types syntactically within another:

```rust
existential type Foo: Deref<Target = [u8; {
    existential type Foo: Debug;
    fn dummy() -> Foo {}
    42
}]>;
```

Reply:
Yep, good point this. It wasn't affecting my tests, but this seems like the correct way to do it.